---
title: "Analysis when there are multiple genes and phenotypes afftecting fitness"
author: "Clare Horscroft"
date: "16/03/2021"
output: html_document
---

Load packages
```{r}
require(systemfit)
require(TwoSampleMR)
require(ggdag)
require(dagitty)
```

# Two phenotypes and one gene

Read in the data
```{r}
setwd("~/Documents/MR-and-selection/Simulations/children/two_pheno_one_gene/")
dat<-read.table("dat.txt",header = TRUE)
```

Look at the phenotypes for the genotypes
```{r}
tapply(dat$pheno1,dat$geno,mean)
tapply(dat$pheno2,dat$geno,mean)
```

Check the fitness of the individuals is as expected (0.02 higher for each allele)

```{r}
tapply(log(dat$fit),dat$geno,mean)
```

Look at the mean number of children per genotype
```{r}
tapply(dat$children,dat$geno,mean)
```

## MR analysis

Using the systemfit method on fitness
```{r}
sf1<-systemfit(log(fit)~pheno1+pheno2,inst = ~geno,method = "2SLS",data=dat)
summary(sf1)
```

Using glm using the number of children
```{r}
mod1<-summary(lm(dat$pheno1~dat$geno))
bgx<-mod1$coefficients[2,1]
segx<-mod1$coefficients[2,2]
mod2<-summary(glm(dat$children~dat$geno,family=poisson("log")))
bgy<-mod2$coefficients[2,1]
segy<-mod2$coefficients[2,2]
mr_wald_ratio(bgx,bgy,segx,segy)
```

# Two phenotypes and five genes

All five genes contribute to the main phenotype, but one gene affects a second phenotype.
Both phenotypes impact fitness.
For this scenario the DAG looks like this:
```{r}
dag <- dagitty('dag{
              SNP1 [pos="0,5"]
              SNP2 [pos="0,4"]
              SNP3 [pos="0,3"]  
              SNP4 [pos="0,2"]
              SNP5 [pos="0,1"]
              Trait1 [pos="1,4"]
              Trait2 [pos="1,2"]
              Fitness  [pos="2,3"]
              SNP1 -> Trait1 -> Fitness
              SNP2 -> Trait1 -> Fitness
              SNP3 -> Trait1 -> Fitness
              SNP4 -> Trait1 -> Fitness
              SNP5 -> Trait1 -> Fitness
              SNP5 -> Trait2 -> Fitness
}')
ggdag(dag, layout = "circle",node_size = 20) +
  theme_void()
```

```{r}
setwd("~/Documents/MR-and-selection/Simulations/children/two_pheno_5_gene/")
dat<-read.table("dat.txt",header = TRUE)
```

Look at the phenotypes for the genotypes
```{r}
tapply(dat$pheno1,dat$geno1,mean)
tapply(dat$pheno2,dat$geno5,mean)
```

## MR analysis

Using the systemfit method on fitness
```{r}
sf1<-systemfit(log(fit)~pheno1+pheno2,inst = ~geno1+geno2+geno3+geno4+geno5,method = "2SLS",data=dat)
summary(sf1)
```

Using the mr function
```{r}
exposure<-t(sapply(2:6,function(x)summary(lm(dat$pheno1~dat[,x]))$coefficients[2,c(1,2)]))
outcome<-t(sapply(2:6,function(x)summary(lm(log(dat$fit)~dat[,x]))$coefficients[2,c(1,2)]))
dat_pheno1<-data.frame(SNP=1:5,id.exposure="pheno1",exposure="pheno1",id.outcome="fit",outcome="fit",beta.exposure=exposure[,1],se.exposure=exposure[,2],beta.outcome=outcome[,1],se.outcome=outcome[,2],mr_keep=TRUE)
dat_pheno1$beta.outcome[dat_pheno1$beta.exposure<0]<--dat_pheno1$beta.outcome[dat_pheno1$beta.exposure<0]
dat_pheno1$beta.exposure[dat_pheno1$beta.exposure<0]<--dat_pheno1$beta.exposure[dat_pheno1$beta.exposure<0]
mr_results <- mr(dat_pheno1, method_list=c("mr_ivw","mr_egger_regression","mr_weighted_median", "mr_weighted_mode"))
mr_results
```

Results return 0.05 as expected

Sensitivity analysis shows that gene 5 is a clear outlier, and is probably pleiotropic:
```{r}
mr_scatter_plot(mr_results, dat_pheno1)
mr_forest_plot(mr_singlesnp(dat_pheno1))
```


## Redo but with number of children as the outcome

```{r}
exposure<-t(sapply(2:6,function(x)summary(lm(dat$pheno1~dat[,x]))$coefficients[2,c(1,2)]))
outcome<-t(sapply(2:6,function(x)summary(glm(dat$children~dat[,x],family = poisson()))$coefficients[2,c(1,2)]))
dat_pheno1<-data.frame(SNP=1:5,id.exposure="pheno1",exposure="pheno1",id.outcome="child",outcome="child",beta.exposure=exposure[,1],se.exposure=exposure[,2],beta.outcome=outcome[,1],se.outcome=outcome[,2],mr_keep=TRUE)
dat_pheno1$beta.outcome[dat_pheno1$beta.exposure<0]<--dat_pheno1$beta.outcome[dat_pheno1$beta.exposure<0]
dat_pheno1$beta.exposure[dat_pheno1$beta.exposure<0]<--dat_pheno1$beta.exposure[dat_pheno1$beta.exposure<0]
mr_results <- mr(dat_pheno1, method_list=c("mr_ivw","mr_egger_regression","mr_weighted_median", "mr_weighted_mode"))
mr_results
```

```{r}
mr_scatter_plot(mr_results, dat_pheno1)
mr_forest_plot(mr_singlesnp(dat_pheno1))
```